content/blog/2019-10-04-InstructionScheduling.md

```diff
@@ -0,0 +1,143 @@
+
+Lookaheads would definitely improve scheduling, but that also significantly increases worst-case complexity. Instead, they use three concrete heuristics that evaluate the candidates' static local properties. In order of importance, they are as follows:
+ 1. Whether an instruction interlocks with any of its immediate successors.
+ 2. The number of immediate successors.
+ 3. The length of the longest path from the instruction to the leaves.
+
+These criteria yields instructions which:
+1. May cause interlocks. This is desirable because it allows instructions that are likely to interlock to be scheduled as early as possible as that gives the greatest number of candidates for subsequent instructions.
+2. Uncover the most potential successors, thereby giving greater freedom of future choices.
+3. Balance the progress along the paths of the dag, which ensures a more even number of choices throughout the process.
+
+The steps in this algorithm for dag traversal and scheduling are outlined pretty clearly in the paper and so we will not duplicate them here. The final result is displayed in the table above. The original order had four interlocks (referring to instructions via line number): 3-4, 5-6, 7-8. and 8-9. The reordered version only has one: 8-1. 
+
+
+### Computational Complexity of the Algorithm
+Let the number of instructions in the basic block be n.
+#### Constructing the dag
+In the worst case, every instruction needs to be compared with the instructions already in dag. During this process, we can also compute the information needed by the heuristic evaluation, such as the length of the longest path and the number of immediate successors. It is thus O(n<sup>2</sup>).
+#### Scheduling the instructions
+To schedule an instruction, we need to evaluate all the candidates based on the heuristic. Evaluation can be done in O(1) because we already build the dag contains all the relevant information. Thus, scheduling is also O(n<sup>2</sup>).
+
+#### Overall complexity
+The overall complexity is O(n<sup>2</sup>). This is significantly better compared to other algorithms in this space which are O(n<sup>4</sup>), even when adjusted to have similar hardware assumptions.
+
+### Experiments
+The authors implemented this instruction scheduler and made the following observations from benchmark results:
+ - In practice, these heuristics effectively remove avoidable interlocks and runs in approximately linear time.
+ - The memory referencing assumptions greatly improve results, and effectiveness increases with better aliasing information (provided by other parts of the compiler).
+ - Carry/Borrow subgraph do not improve much for most programs. Significant improvements only occur when the program is computationally intensive. 
+ - Using more versatile dags proposed by other literature only slightly improves the instruction scheduling effectiveness. 
```

Comment:
Perhaps mention that this is despite those other algorithms having worse asymptotic complexity?